lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -5153,6 +5180,47 @@ ParseResult FIRCircuitParser::parseModule(CircuitOp circuit, bool isPublic,
   return success();
 }
 
+ParseResult FIRCircuitParser::parseFormal(CircuitOp circuit, unsigned indent) {
+  consumeToken(FIRToken::kw_formal);
+  StringRef id, kSpelling;
+  int64_t k = -1;
+  LocWithInfo info(getToken().getLoc(), this);
+
+  // Parse the formal operation
+  if (parseId(id, "expected a formal test name") ||
+      parseToken(FIRToken::comma, "expected ','") ||
+      parseGetSpelling(kSpelling) ||
+      parseToken(FIRToken::identifier, "expected parameter 'k' after ','") ||
+      parseToken(FIRToken::equal, "expected '=' after 'k'") ||
+      parseIntLit(k, "expected integer in k specification") ||
+      parseToken(FIRToken::colon,
+                 "expected ':' after option group definition") ||
+      info.parseOptionalInfo())
+    return failure();
+
+  // Check that the parameter is valid
+  if (kSpelling != "k" || k <= 0)
+    return emitError("Invalid parameter given to formal test: ") << kSpelling,
+           failure();
+
+  // Build out the firrtl mlir op
+  auto builder = circuit.getBodyBuilder();
+  auto formal = builder.create<firrtl::FormalOp>(info.getLoc(), id, k);
+  auto &body = formal.getBody().emplaceBlock();
+
+  // Create the statement parser
+  hw::InnerSymbolNamespace ns(circuit);
+  SymbolTable circuitSymTbl(circuit);
```

Comment:
SymbolTable computation walks entire top-level modules so it would eventually cause O(N^2) regression. Can you defer the actual parsing in a similar way to [Module parsing ](https://github.com/llvm/circt/blob/10180d3ca24f644dd6ef33868c1225a503ff57b1/lib/Dialect/FIRRTL/Import/FIRParser.cpp#L5569-L5577)? Basically we might want to separate the parser into `parseFormalBody` and `parseFormal`